### PR TITLE
doc: add the various storageclass configuration howto

### PR DESCRIPTION
**Pull request by Vicente-Cheng**

### CODE REVIEW

File: docs/rancher/csi-driver-support-various-sc.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  - Rancher Integration
+  - CSI Driver
+  - StorageClass
+---
+
+After the Harvester CSI driver v0.1.15, you can create PVC based on the different StorageClass of Harvester.
+
+## Preparation
+
+It would be best if you did some preparation to support this feature entirely.
```
Comment (guangbochen) [inline docs/rancher/csi-driver-support-various-sc.md:17]: Can you please elaborate on why this pre-configuration is needed here? 
e.g., in order to allow users to balabala, it is recommended to...

File: docs/rancher/csi-driver-support-various-sc.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+- kind: ServiceAccount
+  name: <serviceaccount name>
+  namespace: <namespace>
+  ```
+
+:::note
+
+Please complete these preparations on the Harvester Cluster.
+
+Also, the `serviceaccount name` and `namespace` should be the same as your could provider.
```
Comment (guangbochen) [inline docs/rancher/csi-driver-support-various-sc.md:60]: can u please provide a guideline for how users could allocate them accordingly?

File: docs/rancher/csi-driver-support-various-sc.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+:::
+
+## Deploying
+
+Now, try to create a new StorageClass. You can refer to the [StorageClasses](https://docs.harvesterhci.io/dev/advanced/storageclass) to see how to do it.
```
Comment (guangbochen) [inline docs/rancher/csi-driver-support-various-sc.md:66]:
```suggestion
Now, let's try to create a new StorageClass that you like to use in your guest k8s cluster. You can refer to the [StorageClasses](https://docs.harvesterhci.io/dev/advanced/storageclass) for more details.
```

File: docs/rancher/csi-driver-support-various-sc.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+:::note
+
+The `Provisioner` should choose `Harvester`.
+
+The parameter `Host StorageClass` should be the StorageClass created on the Harvester Cluster.
+
+:::
+
+Then, you could use this new StorageClass to create the PVC based on it.
```
Comment (guangbochen) [inline docs/rancher/csi-driver-support-various-sc.md:84]:
```suggestion
You can then use this new `StorageClass` to create a PVC based on it, which will use the above-defined `Host StorageClass` to provision volumes on the bare-metal cluster.
```

File: docs/rancher/csi-driver-support-various-sc.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+sidebar_position: 6
+sidebar_label: Various StorageClass of Harvester CSI Driver 
+title: "Various StorageClass of Harvester CSI Driver"
+keywords:
+  - Harvester
+  - harvester
+  - Rancher Integration
+  - CSI Driver
+  - StorageClass
+---
```
[the remark this answers is not in the document]
Comment (Vicente-Cheng) [reply docs/rancher/csi-driver-support-various-sc.md]: Updated

File: docs/rancher/csi-driver-support-various-sc.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+After the Harvester CSI driver v0.1.15, you can create PVC based on the different StorageClass of Harvester.
```
[the remark this answers is not in the document]
Comment (Vicente-Cheng) [reply docs/rancher/csi-driver-support-various-sc.md]: Sure, updated!

File: docs/rancher/csi-driver-support-various-sc.md (hunk 1 above)
Comment (Vicente-Cheng) [reply docs/rancher/csi-driver-support-various-sc.md]: Updated!

File: docs/rancher/csi-driver-support-various-sc.md (hunk 2 above)
Comment (Vicente-Cheng) [reply docs/rancher/csi-driver-support-various-sc.md]: Updated! Please help to review the latest version! Thanks

File: docs/rancher/csi-driver.md [hunk 7]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
 - Install `Harvester CSI Driver` from the Rancher marketplace.
 
   ![](/img/v1.2/rancher/install-harvester-csi-driver-in-k3s.png)
+
+### Various StorageClass Supported
+
+After the Harvester CSI driver v0.1.15 (check with the following `note` to see the detailed version of rke2), you can create PVC based on the different StorageClass of Harvester.
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:105]:
```suggestion
Starting with Harvester CSI driver v0.1.15, you can create a PersistentVolumeClaim (PVC) based on a different StorageClass.
```

File: docs/rancher/csi-driver.md [hunk 8]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+:::note
+
+Supported CSI driver version: >= v0.1.15
+
+Supported RKE2 version: >= v1.23.16+rke2r1, >= v1.24.10+rke2r1, >= v1.25.6+rke2r1, >= v1.26.1+rke2r1, >= v1.27.1+rke2r1, and later version
+
+:::
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:113]:
```suggestion
The following RKE2 versions are supported for Harvester CSI driver v0.1.15 and later:
- v1.23.16+rke2r1 and later
- v1.24.10+rke2r1 and later
- v1.25.6+rke2r1 and later
- v1.26.1+rke2r1 and later
- v1.27.1+rke2r1 and later
```

File: docs/rancher/csi-driver.md [hunk 9]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+#### Preparation
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:115]:
```suggestion
#### Prerequisites
```

File: docs/rancher/csi-driver.md [hunk 10]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+CSI driver needs `RBAC` to expose the error message when creating PVC with the non-existence StorageClass (as shown in the following figure).
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:117]:
```suggestion
Please complete these perquisites on the Harvester Cluster. The Harvester CSI driver needs **RBAC** to expose error messages. This is important for displaying an error message when creating a PVC with a StorageClass that does not exist, as shown in the following figure.
```

File: docs/rancher/csi-driver.md [hunk 11]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+![](/img/v1.2/rancher/error_event_csi_driver.png)
+
+
+We need to do the following `RBAC` setup, or you can not see any error message even creating PVC with the non-existence StorageClass (Because we do the compatibility with the previous version).
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:122]:
```suggestion
Perform the following steps to setup **RBAC** to enable seeing error messages.
```

File: docs/rancher/csi-driver.md [hunk 12]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+  - storage.k8s.io
+  resources:
+  - storageclasses
+  verbs:
+  - get
+  - list
+  - watch
+```
+
+1. Then create the `clusterrolebinding` to associate with this. You can refer to the following manifest.
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:145]:
```suggestion
1. Then create the `clusterrolebinding` to associate with  the new `clusterrole` with the following manifest.
```

File: docs/rancher/csi-driver.md [hunk 13]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+:::note
+
+Please complete these preparations on the Harvester Cluster.
+
+Also, the `serviceaccount name` and `namespace` should be the same as your could provider.
+
+:::
+
+We could check with the following steps to confirm the serviceaccount name and namespace on the cloud provider
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:169]:
```suggestion
Make sure the `serviceaccount name` and `namespace` are the same as your cloud provider. Perform the following steps to see the `serviceaccount name` and `namespace` for your cloud provider.
```

File: docs/rancher/csi-driver.md [hunk 14]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+1. Found the cloudprovider rolebinding.
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:171]:
```suggestion
1. Find the `rolebinding` for your cloud provider .
```

File: docs/rancher/csi-driver.md [hunk 15]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+```
+# kubectl get rolebinding -A |grep harvesterhci.io:cloudprovider
+default                                 default-rke2-guest-01                                ClusterRole/harvesterhci.io:cloudprovider             7d1h
+```
+
+2. Get the subjects info on this rolebinding.
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:178]:
```suggestion
2. Get the `subjects` info on this `rolebinding`.
```

File: docs/rancher/csi-driver.md [hunk 16]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+```
+kubectl get rolebinding default-rke2-guest-01 -n default -o yaml |yq -e '.subjects'
+```
+
+3. You can find the ServiceAccount info as follows:
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:183]:
```suggestion
3. Find the `ServiceAccount` info as follows:
```

File: docs/rancher/csi-driver.md [hunk 17]
```diff
@@ -99,3 +99,112 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+:::note
+
+The `Provisioner` should choose `Harvester`.
+
+The parameter `Host StorageClass` should be the StorageClass created on the Harvester Cluster.
+
+:::
+
+You can then use this new `StorageClass` to create a PVC based on it, which will use the above-defined `Host StorageClass` to provision volumes on the bare-metal cluster.
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:210]:
```suggestion
1. Create a new StorageClass that you would like to use in your guest k8s cluster. You can refer to the [StorageClasses](https://docs.harvesterhci.io/dev/advanced/storageclass) for more details.

  As show in the following figure, create a new StorageClass named **replica-2**.

  ![](/img/v1.2/rancher/sc-replica-2.png)

  For example, create a new StorageClass on the downstream cluster as follows associated with the StorageClass created on the Harvester Cluster named **replica-2**.

  ![](/img/v1.2/rancher/downstream-cluster-sc-creation.png)

  :::note

  When choosing a **Provisioner**, select **Harvester (CSI)**. The parameter **Host StorageClass** should be the StorageClass created on the Harvester Cluster.

  :::
  
1. You can create a PVC based on this new **StorageClass**, which uses the **Host StorageClass** to provision volumes on the bare-metal cluster.
```

**Review by LucasSaintarbor — COMMENTED** [review]
@Vicente-Cheng Made some changes to docs/rancher. 

- Bolding UI elements
- Reducing the amount of **Notes** on a page. 

If these changes are okay with you, I'll add them to version-v1.1/rancher/docs as well.

**Comment by Vicente-Cheng** [comment]
> @Vicente-Cheng Made some changes to docs/rancher.
> 
> * Bolding UI elements
> * Reducing the amount of **Notes** on a page.
> 
> If these changes are okay with you, I'll add them to version-v1.1/rancher/docs as well.

Hi @LucasSaintarbor, Thanks for your review.
I accept all a suggestion with a batch and copy them to version-v1.1/rancher/docs.
Could you help recheck it?  Thanks!

File: docs/rancher/csi-driver.md [hunk 18]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+  - storage.k8s.io
+  resources:
+  - storageclasses
+  verbs:
+  - get
+  - list
+  - watch
+```
+
+1. Then create the `clusterrolebinding` to associate with  the new `clusterrole` with the following manifest.
```
Comment (bk201) [inline docs/rancher/csi-driver.md:144]: The number list on the preview page needs to be corrected. It should be `2.` but shown as `1.`.

**Review by bk201 — APPROVED** [review]

File: docs/rancher/csi-driver.md [hunk 19]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
 - Install `Harvester CSI Driver` from the Rancher marketplace.
 
   ![](/img/v1.2/rancher/install-harvester-csi-driver-in-k3s.png)
+
+### Various StorageClass Supported
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:103]:
```suggestion
## Various StorageClass Supported
```

File: docs/rancher/csi-driver.md [hunk 20]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+Starting with Harvester CSI driver v0.1.15, you can create a PersistentVolumeClaim (PVC) based on a different StorageClass.
+
+The following RKE2 versions are supported for Harvester CSI driver v0.1.15 and later:
+- v1.23.16+rke2r1 and later
+- v1.24.10+rke2r1 and later
+- v1.25.6+rke2r1 and later
+- v1.26.1+rke2r1 and later
+- v1.27.1+rke2r1 and later
+
+#### Prerequisites
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:114]:
```suggestion
### Prerequisites
```

File: docs/rancher/csi-driver.md [hunk 21]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+```
+
+3. Find the `ServiceAccount` info as follows:
+```
+- kind: ServiceAccount
+  name: rke2-guest-01
+  namespace: default
+```
+
+#### Deploying
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:181]:
```suggestion
### Deploying
```

File: versioned_docs/version-v1.1/rancher/csi-driver.md [hunk 22]
```diff
@@ -98,3 +98,101 @@ The Harvester CSI driver should now be deployed successfully.
 - Install `Harvester CSI Driver` from the Rancher marketplace.
 
   ![](/img/v1.1/rancher/install-harvester-csi-driver-in-k3s.png)
+
+### Various StorageClass Supported
```
Comment (LucasSaintarbor) [inline versioned_docs/version-v1.1/rancher/csi-driver.md:102]:
```suggestion
## Various StorageClass Supported
```

File: versioned_docs/version-v1.1/rancher/csi-driver.md [hunk 23]
```diff
@@ -98,3 +98,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+Starting with Harvester CSI driver v0.1.15, you can create a PersistentVolumeClaim (PVC) based on a different StorageClass.
+
+The following RKE2 versions are supported for Harvester CSI driver v0.1.15 and later:
+- v1.23.16+rke2r1 and later
+- v1.24.10+rke2r1 and later
+- v1.25.6+rke2r1 and later
+- v1.26.1+rke2r1 and later
+- v1.27.1+rke2r1 and later
+
+#### Prerequisites
```
Comment (LucasSaintarbor) [inline versioned_docs/version-v1.1/rancher/csi-driver.md:113]:
```suggestion
### Prerequisites
```

File: versioned_docs/version-v1.1/rancher/csi-driver.md [hunk 24]
```diff
@@ -98,3 +98,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+```
+
+3. Find the `ServiceAccount` info as follows:
+```
+- kind: ServiceAccount
+  name: rke2-guest-01
+  namespace: default
+```
+
+#### Deploying
```
Comment (LucasSaintarbor) [inline versioned_docs/version-v1.1/rancher/csi-driver.md:180]:
```suggestion
### Deploying
```

File: versioned_docs/version-v1.1/rancher/csi-driver.md [hunk 25]
```diff
@@ -98,3 +98,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+subjects:
+- kind: ServiceAccount
+  name: <serviceaccount name>
+  namespace: <namespace>
+  ```
+
+Make sure the `serviceaccount name` and `namespace` are the same as your cloud provider. Perform the following steps to see the `serviceaccount name` and `namespace` for your cloud provider.
+
+1. Find the `rolebinding` for your cloud provider .
+
```
Comment (LucasSaintarbor) [inline versioned_docs/version-v1.1/rancher/csi-driver.md:162]:
```suggestion
```

File: docs/rancher/csi-driver.md [hunk 26]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+  - storageclasses
+  verbs:
+  - get
+  - list
+  - watch
+```
+
+2. Then create the `clusterrolebinding` to associate with  the new `clusterrole` with the following manifest.
+```
+  apiVersion: rbac.authorization.k8s.io/v1
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:146]:
```suggestion
 apiVersion: rbac.authorization.k8s.io/v1
```

File: docs/rancher/csi-driver.md [hunk 27]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+  name: <namespace>-<serviceaccount name>
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: harvesterhci.io:csi-driver
+subjects:
+- kind: ServiceAccount
+  name: <serviceaccount name>
+  namespace: <namespace>
+  ```
```
Comment (LucasSaintarbor) [inline docs/rancher/csi-driver.md:158]:
```suggestion
```
```

File: versioned_docs/version-v1.1/rancher/csi-driver.md [hunk 28]
```diff
@@ -98,3 +98,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+  - storageclasses
+  verbs:
+  - get
+  - list
+  - watch
+```
+
+2. Then create the `clusterrolebinding` to associate with  the new `clusterrole` with the following manifest.
+```
+  apiVersion: rbac.authorization.k8s.io/v1
```
Comment (LucasSaintarbor) [inline versioned_docs/version-v1.1/rancher/csi-driver.md:145]:
```suggestion
apiVersion: rbac.authorization.k8s.io/v1
```

File: versioned_docs/version-v1.1/rancher/csi-driver.md [hunk 29]
```diff
@@ -98,3 +98,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+  name: <namespace>-<serviceaccount name>
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: harvesterhci.io:csi-driver
+subjects:
+- kind: ServiceAccount
+  name: <serviceaccount name>
+  namespace: <namespace>
+  ```
```
Comment (LucasSaintarbor) [inline versioned_docs/version-v1.1/rancher/csi-driver.md:157]:
```suggestion
```
```

**Review by LucasSaintarbor — APPROVED** [review]
Made a few small style changes
- Changed headers from H3 > H2 and H4 > H3
- Removed extra white space in code blocks
- Fixed code block

LGTM 👍

**Comment by LucasSaintarbor** [comment]
> > @Vicente-Cheng Made some changes to docs/rancher.
> > 
> > * Bolding UI elements
> > * Reducing the amount of **Notes** on a page.
> > 
> > If these changes are okay with you, I'll add them to version-v1.1/rancher/docs as well.
> 
> Hi @LucasSaintarbor, Thanks for your review. I accept all a suggestion with a batch and copy them to version-v1.1/rancher/docs. Could you help recheck it? Thanks!

Thanks @Vicente-Cheng!! Made a few small style changes.

**Comment by Vicente-Cheng** [comment]
> Made a few small style changes
> 
> * Changed headers from H3 > H2 and H4 > H3
> * Removed extra white space in code blocks
> * Fixed code block
> 
> LGTM 👍

Thanks, @LucasSaintarbor. All suggestions are updated!

File: docs/rancher/csi-driver.md [hunk 30]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
 - Install `Harvester CSI Driver` from the Rancher marketplace.
 
   ![](/img/v1.2/rancher/install-harvester-csi-driver-in-k3s.png)
+
+## Various StorageClass Supported
```
Comment (guangbochen) [inline docs/rancher/csi-driver.md:103]:
```suggestion
## Passthrough Custom StorageClass
```

File: docs/rancher/csi-driver.md [hunk 31]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+Starting with Harvester CSI driver v0.1.15, you can create a PersistentVolumeClaim (PVC) based on a different StorageClass.
+
+The following RKE2 versions are supported for Harvester CSI driver v0.1.15 and later:
```
Comment (guangbochen) [inline docs/rancher/csi-driver.md:107]:
```suggestion
Harvester CSI driver v0.1.15 is supported out of the box starting with the following RKE2 versions; for RKE1 you need to manually install the CSI driver chart:
```

File: docs/rancher/csi-driver.md [hunk 32]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+- v1.23.16+rke2r1 and later
+- v1.24.10+rke2r1 and later
+- v1.25.6+rke2r1 and later
+- v1.26.1+rke2r1 and later
+- v1.27.1+rke2r1 and later
+
+### Prerequisites
+
+Please complete these perquisites on the Harvester Cluster. The Harvester CSI driver needs **RBAC** to expose error messages. This is important for displaying an error message when creating a PVC with a StorageClass that does not exist, as shown in the following figure.
```
Comment (guangbochen) [inline docs/rancher/csi-driver.md:116]:
```suggestion
Please add the following additional perquisites to your Harvester cluster. The Harvester CSI driver requires proper **RBAC** to display error messages. This is important for displaying an error message when creating a PVC with a StorageClass that does not exist, as shown in the following figure.
```

File: docs/rancher/csi-driver.md [hunk 33]
```diff
@@ -99,3 +99,101 @@ The Harvester CSI driver should now be deployed successfully.
...
+
+  ![](/img/v1.2/rancher/downstream-cluster-sc-creation.png)
+
+  :::note
+
+  When choosing a **Provisioner**, select **Harvester (CSI)**. The parameter **Host StorageClass** should be the StorageClass created on the Harvester Cluster.
+
+  :::
+  
+1. You can create a PVC based on this new **StorageClass**, which uses the **Host StorageClass** to provision volumes on the bare-metal cluster.
```
Comment (guangbochen) [inline docs/rancher/csi-driver.md:199]:
```suggestion
1. You can now create a PVC based on this new **StorageClass**, which uses the **Host StorageClass** to provision volumes on the bare-metal cluster.
```

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Pull request merged**